### PR TITLE
[WIP] Diffusers

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# WebUI meets diffusers

For now, very basic txt2image works:

```
./webui.sh --share --diffusers
```

make sure to have the following settings in the `config.json`:

```
"add_model_hash_to_info": false,                                                                                                                                                               "add_model_name_to_info": false,
"disable_all_extensions": "all",
```

### CODE REVIEW

File: modules/processing.py [hunk 1]
```diff
@@ -671,60 +683,78 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
...
-                x_sample = 255. * np.moveaxis(x_sample.cpu().numpy(), 0, 2)
-                x_sample = x_sample.astype(np.uint8)
+                x_sample = (255. * x_sample).astype(np.uint8)
 
                 if p.restore_faces:
                     if opts.save and not p.do_not_save_samples and opts.save_images_before_face_restoration:
                         images.save_image(Image.fromarray(x_sample), p.outpath_samples, "", seeds[i], prompts[i], opts.samples_format, info=infotext(n, i), p=p, suffix="-before-face-restoration")
 
-                    devices.torch_gc()
+                        devices.torch_gc()
```
Comment (patrickvonplaten) [inline modules/processing.py:754]:
```suggestion
                    devices.torch_gc()

```

File: modules/processing.py [hunk 2]
```diff
@@ -671,60 +683,78 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
...
 
-                    x_sample = modules.face_restoration.restore_faces(x_sample)
-                    devices.torch_gc()
+                        x_sample = modules.face_restoration.restore_faces(x_sample)
```
Comment (patrickvonplaten) [inline modules/processing.py:756]:
```suggestion
                    x_sample = modules.face_restoration.restore_faces(x_sample)

```

File: modules/processing.py [hunk 3]
```diff
@@ -671,60 +683,78 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
...
+                        devices.torch_gc()
```
Comment (patrickvonplaten) [inline modules/processing.py:757]:
```suggestion
                    devices.torch_gc()

```

File: modules/processing.py [hunk 4]
```diff
@@ -766,11 +796,11 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
                     if opts.return_mask_composite:
                         output_images.append(image_mask_composite)
 
-            del x_samples_ddim
+                del x_samples_ddim
```
Comment (patrickvonplaten) [inline modules/processing.py:799]:
```suggestion
            del x_samples_ddim

```

File: modules/processing.py [hunk 5]
```diff
@@ -766,11 +796,11 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
                     if opts.return_mask_composite:
                         output_images.append(image_mask_composite)
 
-            del x_samples_ddim
+                del x_samples_ddim
 
-            devices.torch_gc()
+                devices.torch_gc()
```
Comment (patrickvonplaten) [inline modules/processing.py:801]:
```suggestion
            devices.torch_gc()

```

File: modules/processing.py [hunk 6]
```diff
@@ -766,11 +796,11 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
...
                         output_images.append(image_mask_composite)
 
-            del x_samples_ddim
+                del x_samples_ddim
 
-            devices.torch_gc()
+                devices.torch_gc()
 
-            state.nextjob()
+                state.nextjob()
```
Comment (patrickvonplaten) [inline modules/processing.py:803]:
```suggestion
            state.nextjob()

```

File: modules/txt2img.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import modules.scripts
+from typing import List
```
Comment (patrickvonplaten) [inline modules/txt2img.py:2]:
```suggestion

```

File: modules/txt2img.py [hunk 8]
```diff
@@ -42,9 +43,11 @@ def txt2img(id_task: str, prompt: str, negative_prompt: str, prompt_styles, step
     )
     p.scripts = modules.scripts.scripts_txt2img
     p.script_args = args
+
```
Comment (patrickvonplaten) [inline modules/txt2img.py:46]:
```suggestion

```

File: modules/txt2img.py [hunk 9]
```diff
@@ -42,9 +43,11 @@ def txt2img(id_task: str, prompt: str, negative_prompt: str, prompt_styles, step
     )
     p.scripts = modules.scripts.scripts_txt2img
     p.script_args = args
+
     processed = modules.scripts.scripts_txt2img.run(p, *args)
     if processed is None:
         processed = process_images(p)
+
```
Comment (patrickvonplaten) [inline modules/txt2img.py:50]:
```suggestion

```

File: webui.py [hunk 10]
```diff
@@ -9,8 +9,11 @@
 from rich import print # pylint: disable=W0622
 from modules import timer, errors
 
+from diffusers import DiffusionPipeline, UniPCMultistepScheduler
+
 startup_timer = timer.Timer()
 
+
```
Comment (patrickvonplaten) [inline webui.py:16]:
```suggestion

```

File: webui.py [hunk 11]
```diff
@@ -147,17 +150,27 @@ def sigint_handler(_sig, _frame):
...
+            print("Stable diffusion model failed to load")
+            exit(1)
+        if shared.sd_model is None:
+            print("No stable diffusion model loaded")
+            exit(1)
+        shared.opts.data["sd_model_checkpoint"] = shared.sd_model.sd_checkpoint_info.title
+        shared.opts.onchange("sd_model_checkpoint", wrap_queued_call(lambda: modules.sd_models.reload_model_weights()))
+    else:
+        print("Loading `diffusers` backend")
+        model_id = "runwayml/stable-diffusion-v1-5"
```
Comment (patrickvonplaten) [inline webui.py:167]: For now hardcoded

File: modules/processing.py [hunk 12]
```diff
@@ -671,61 +683,82 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
...
-                x_sample = x_sample.astype(np.uint8)
+                x_sample = (255. * x_sample).astype(np.uint8)
 
                 if p.restore_faces:
                     if opts.save and not p.do_not_save_samples and opts.save_images_before_face_restoration:
                         images.save_image(Image.fromarray(x_sample), p.outpath_samples, "", seeds[i], prompts[i], opts.samples_format, info=infotext(n, i), p=p, suffix="-before-face-restoration")
 
                     devices.torch_gc()
 
+
```
Comment (patrickvonplaten) [inline modules/processing.py:756]:
```suggestion

```

File: modules/processing.py [hunk 13]
```diff
@@ -671,61 +683,82 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
...
                     x_sample = modules.face_restoration.restore_faces(x_sample)
+
```
Comment (patrickvonplaten) [inline modules/processing.py:758]:
```suggestion

```

File: modules/processing.py [hunk 14]
```diff
@@ -768,10 +801,13 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
 
             del x_samples_ddim
 
+
```
Comment (patrickvonplaten) [inline modules/processing.py:804]:
```suggestion

```

File: modules/processing.py [hunk 15]
```diff
@@ -768,10 +801,13 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
 
             del x_samples_ddim
 
+
             devices.torch_gc()
 
+
```
Comment (patrickvonplaten) [inline modules/processing.py:807]:
```suggestion

```

File: modules/processing.py [hunk 16]
```diff
@@ -768,10 +801,13 @@ def get_conds_with_caching(function, required_prompts, steps, cache):
 
             del x_samples_ddim
 
+
             devices.torch_gc()
 
+
             state.nextjob()
 
+
```
Comment (patrickvonplaten) [inline modules/processing.py:810]:
```suggestion

```

**Comment by ilikenwf** [comment]
Is this for performance with compiling models to a tensor compatible format, or is it something else? Pardon my ignorance.

**Comment by dsully** [comment]
Happy to see this!

**Comment by vladmandic** [comment]
closing the pr as manual merge to master is complete and further work should be tracked via <https://github.com/users/vladmandic/projects/1>

**Pull request closed without merging**